Sprint 17

Lisiykod opened the pull request.

Lisiykod committed:
добавлена базовая локализация для онбординга и главного экрана с трекерами

Lisiykod committed:
добавлена плюрализация строк для количества дней выполненных трекеров

Lisiykod committed:
реализован метод удаления трекера

Lisiykod committed:
реализована функция закрепления трекера

Lisiykod committed:
реализована функция открепления трекера + скорректирован метод закрепления

Lisiykod committed:
реализована функция редактирования трекера и нерегулярного события

Lisiykod committed:
реализована функция поиска по трекерам

Lisiykod committed:
реализована функция фильтрации по кнопке

Lisiykod committed:
реализован показ статистики по завершенным трекерам

Lisiykod committed:
добавлен значок пина для закрепленных трекеров

Lisiykod committed:
добавлены переводы для главного экрана и статистики

Lisiykod committed:
реализована функция редактирования и удаления категории

Lisiykod committed:
скорректирована логика сохранения трекера в режиме редактирования при изменении категории,
добавлено сохранение отображения выбранной категории,
скорректирована логика отображения выбранного фильтра (отображения галочки)

Lisiykod committed:
скорректировано отображение эмодзи и цвета трекера при редактировании

Lisiykod committed:
добавлены скриншот тесты

Lisiykod committed:
добавлена темная тема

Lisiykod committed:
добавлена метрика
+ устранен баг доступности кнопки "Создать" при создании трекера и нерегулярного события

Lisiykod committed:
в коммит добавлены pod файлы

Lisiykod committed:
в коммит добавлены файлы workspace и snapshots

Lisiykod committed:
скорректировано отображение экрана онбординга в темном режиме

Lisiykod committed:
устранен баг с некорректным отображением статистики

Lisiykod committed:
скорректированы имена событий аналитики

Lisiykod committed:
скорректировано отображение цвета названия трекера для темной темы

Lisiykod committed:
скорректировано отображение таб бара для светлой и темной темы

Helvetii951 commented on Podfile.lock:
Очень желательно убрать файлы, которые генерит CocoaPods:

- Podfile.lock

Нужно их занести в .gitignore.

Helvetii951 commented on Tracker/AppDelegate.swift:
```diff
@@ -6,12 +6,18 @@
+        
+        guard let configuration = AppMetricaConfiguration(apiKey: "a1bd30a8-884f-4fa8-9df2-2eadecee9db1") else { return true }
+        AppMetrica.activate(with: configuration)
+        
```
Рекомендую занести ключ в [enum](https://www.swiftbysundell.com/articles/powerful-ways-to-use-swift-enums/)

Helvetii951 commented on Tracker/Onboarding/OnboardingPageViewController.swift:
```diff
@@ -10,13 +10,16 @@ import UIKit
 final class OnboardingPageViewController: UIPageViewController {
     
     private lazy var pages: [UIViewController] = {
+        let firstOnboardingTitle = NSLocalizedString("firstOnboardingTitle", comment: "Text displayed on first onboarding page")
+        let secondOnboardingTitle = NSLocalizedString("secondOnboardingTitle", comment: "Text displayed on second onboarding page")
+        
```
Неплохое решение поместить цвета в экстншн, но есть более надежные и универсальные подходы:
1) Использовать библиотеки R.Swift или SwiftGen.
2) Генерить ресурсы с помощью xCode.

В[ этой статье](https://sarunw.com/posts/swift-symbols-for-asset-catalog/) описаны оба.

Helvetii951 commented on Tracker/Resources/ru.lproj/Localizable.strings:
```diff
@@ -0,0 +1,45 @@
+"emptyTrackersLabel" = "Что будем отслеживать?";
+"emptySearchOrFilterLabel" = "Ничего не найдено";
+
+"pinTitle" = "Закрепить";
+"unpinTitile" = "Открепить";
+"deleteTitile" = "Удалить";
+"editTitile" = "Редактировать";
+"deleteActionSheetTitle" = "Уверены что хотите удалить трекер?";
+"deleteActionTitle" = "Удалить";
+"cancelActionTitle" = "Отмена";
+"searchPlaceholderTitle" = "Поиск";
+"filterButton" = "Фильтры";
```
Рекомендую называть локализованные строки так:
```
"module_name.screen_name.exact_string_name" = "some string";
```

Helvetii951 commented on Tracker/Statistics/StatisticsViewController.swift:
```diff
@@ -32,23 +35,90 @@ final class StatisticsViewController: UIViewController {
     }
     
-    private func setupConstraints() {
+    private func setupVisibleElements() {
+        if completedTrackers.isEmpty {
+            tableView.isHidden = true
+            stackView.isHidden = false
+        } else {
+            stackView.isHidden = true
+            tableView.isHidden = false
+        }
+    }
```
Можно чуть упростить код
```swift
        let comletedTrackerArrayIsEmpty = completedTrackers.isEmpty
        tableView.isHidden = comletedTrackerArrayIsEmpty
        stackView.isHidden = !comletedTrackerArrayIsEmpty
```

Helvetii951 commented on Tracker/Trackers/Presentation/CategoriesListViewController.swift:
```diff
@@ -185,26 +203,82 @@ extension CategoriesListViewController: UITableViewDelegate {
+        
+        let configuration = UIContextMenuConfiguration(identifier: nil, previewProvider: nil) { [weak self] _ in
+            guard let self else { return UIMenu() }
+            
+            let editAction = UIAction(title: "Редактировать") { _ in
+                self.editTracker(at: indexPath)
+            }
+            
+            let deleteAction = UIAction(title: "Удалить", attributes: .destructive) { _ in
+                self.deleteTracker(at: indexPath)
+            }
```
Все же лучше заносить строки сразу в локализацию. Будет в разы проще добваить новые строки, если пондабится.

Helvetii951 commented on Tracker/Trackers/Service/AnalyticsService.swift:
```diff
@@ -0,0 +1,19 @@
+import Foundation
+import AppMetricaCore
+
+final class AnalyticsService {
+    
+    func trackerEvent(name: String, parameters: [AnyHashable: Any]) {
+        AppMetrica.reportEvent(name: name, parameters: parameters) { error in
+            print("REPORT ERROR: %@", error.localizedDescription)
+        }
+    }
+    
```
Я бы занес сюда инициализацию сервиса с ключом в статическую функцию. 

И эту можно сделать статической, так будет потрачено меньше вычислительных ресурсов.

Helvetii951 commented on Tracker/Trackers/Service/TrackersService.swift:
```diff
@@ -86,20 +87,54 @@ final class TrackersService {
+    }
+    
+    func updateTracker(_ tracker: Tracker) {
+//        let category = fetchCategories().compactMap { category -> TrackerCategory in
+//            if category.trackers.contains(where: {$0.id == tracker.id}) {
+//                trackerStore.updateTracker(tracker: tracker)
+//            }
+//        }
+        
+//        fetchCategories().forEach { category in
+//            if category.trackers.contains(where: {$0.id == tracker.id}) {
+//                trackerStore.updateTracker(tracker: tracker)
+//            } else {
+//                addTracker(tracker: tracker, for: <#T##String#>)
+//            }
+//            
+//        }
```
Это последний спринт модуля. Что-то вроде "релиза". Принято очищать код от лишних комментариев.

Helvetii951 commented on TrackerTests/__Snapshots__/TrackerTests/testViewController.1.png:
Лучше не хранить сгенеренные сторонними библиотеками файлы в репозитории, а выносить в .gitignore.

Lisiykod committed:
скорректировано отображение пустого экрана после выбора фильтра +
небольшие улучшения кода после ревью - удаление лишних комментариев, обновлен сервис аналитики, упрощена функция отображения стэков в статистике
